### PR TITLE
Adds hands-free partner import runner 

### LINKED ISSUE #5130: Partner Data Import Bot (ol-home0 Docker Service)
Part of #2034

### Stakeholders
@BharatKalluri

### PR DESCRIPTION
**Pull request by mekarpeles**
- run by olsystem cron
- processes a batch of csv book records
- marshals them to open library json records
- adds them to the ImportBot import_item batch queue
- ImportBot routinely polls db and imports queue records

Closes #5130

### CODE REVIEW

**Comment by mekarpeles** [comment]
This still applies
https://github.com/internetarchive/olsystem/pull/134#discussion_r709322582

File: scripts/partner_batch_imports.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+`import_item` table (http://openlibrary.org/admin/imports)
+which queues items to be imported via the
+Open Library JSON import API: https://openlibrary.org/api/import
+"""
+
+import os
+import re
+import sys
+import web
+import json
```
Comment (mekarpeles) [inline scripts/partner_batch_imports.py:13]:
```suggestion
```

File: scripts/partner_batch_imports.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+import datetime
+from datetime import timedelta
+import logging
+import requests
+
+# Add openlibrary into our path so we can process config + batch functions
```
Comment (mekarpeles) [inline scripts/partner_batch_imports.py:19]: will consider removing after path this PR lands

File: scripts/partner_batch_imports.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+sys.path.insert(0, os.path.abspath(os.path.join(os.sep, 'openlibrary')))
```
Comment (mekarpeles) [inline scripts/partner_batch_imports.py:20]:
```suggestion
from openlibrary.core.imports import Batch
from infogami import config
```

File: scripts/partner_batch_imports.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+from openlibrary.config import load_config
+load_config(
+    os.path.abspath(os.path.join(
+        os.sep, 'olsystem', 'etc', 'openlibrary.yml')))
+from openlibrary.core.imports import Batch
```
Comment (mekarpeles) [inline scripts/partner_batch_imports.py:25]:
```suggestion
```

File: scripts/partner_batch_imports.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+from infogami import config
```
Comment (mekarpeles) [inline scripts/partner_batch_imports.py:26]:
```suggestion
```

File: scripts/partner_batch_imports.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        self.length, self.width, self.height = (
+            data[40], data[41], data[42]
+        )
+
+        # Assert importable
+        assert self.isbn_13
+        for field in self.REQUIRED_FIELDS:
+            assert getattr(self, field)
+
+
```
Comment (mekarpeles) [inline scripts/partner_batch_imports.py:88]:
```suggestion
```

File: scripts/partner_batch_imports.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        contributors = (
+            (data[21+i*3], data[22+i*3], data[23+i*3]) for i in range(5)
+        )
+
+        # form list of author dicts
+        authors = [make_author(c) for c in contributors if c[0]]
+        return authors
+
+
```
Comment (mekarpeles) [inline scripts/partner_batch_imports.py:109]:
```suggestion
```

File: scripts/tests/test_partner_batch_imports.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+from ..partner_batch_imports import Biblio
+
+csv_row = "USA01961304|0962561851||9780962561856|AC|I|TC||B||Sutra on Upasaka Precepts|The||||||||2006|20060531|Heng-ching, Shih|TR||||||||||||||226|ENG||0.545|22.860|15.240|||||||P|||||||74474||||||27181|USD|30.00||||||||||||||||||||||||||||SUTRAS|BUDDHISM_SACRED BOOKS|||||||||REL007030|REL032000|||||||||HRES|HRG|||||||||RB,BIP,MIR,SYN|1961304|00|9780962561856|67499962||PRN|75422798|||||||BDK America||1||||||||10.1604/9780962561856|91-060120||20060531|||||REL007030||||||"
```
Comment (mekarpeles) [inline scripts/tests/test_partner_batch_imports.py:3]:
```suggestion
csv_row = "USA01961304|0962561851||9780962561856|AC|I|TC||B||Sutra on Upasaka Precepts|The||||||||2006|20060531|Heng-ching, Shih|TR||||||||||||||226|ENG||0.545|22.860|15.240|||||||P|||||||74474||||||27181|USD|30.00||||||||||||||||||||||||||||SUTRAS|BUDDHISM_SACRED BOOKS|||||||||REL007030|REL032000|||||||||HRES|HRG|||||||||RB,BIP,MIR,SYN|1961304|00|9780962561856|67499962||PRN|75422798|||||||BDK America||1||||||||10.1604/9780962561856|91-060120||20060531|||||REL007030||||||"  # noqa: E501


```

File: scripts/tests/test_partner_batch_imports.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
```
Comment (mekarpeles) [inline scripts/tests/test_partner_batch_imports.py:4]:
```suggestion


```

File: scripts/tests/test_partner_batch_imports.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
+from ..partner_batch_imports import Biblio
+
+csv_row = "USA01961304|0962561851||9780962561856|AC|I|TC||B||Sutra on Upasaka Precepts|The||||||||2006|20060531|Heng-ching, Shih|TR||||||||||||||226|ENG||0.545|22.860|15.240|||||||P|||||||74474||||||27181|USD|30.00||||||||||||||||||||||||||||SUTRAS|BUDDHISM_SACRED BOOKS|||||||||REL007030|REL032000|||||||||HRES|HRG|||||||||RB,BIP,MIR,SYN|1961304|00|9780962561856|67499962||PRN|75422798|||||||BDK America||1||||||||10.1604/9780962561856|91-060120||20060531|||||REL007030||||||"  # noqa: E501
+
+
+
```
Comment (mekarpeles) [inline scripts/tests/test_partner_batch_imports.py:6]:
```suggestion
```

File: scripts/partner_batch_imports.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+        self.copyright = data[19]
+        self.issn = data[54]
+        self.doi = data[145]
+        self.lccn = data[146]
+        self.dewey = data[49]
+        # physical_dimensions
+        # e.g. "5.4 x 4.7 x 0.2 inches"
+        self.length, self.width, self.height = (
+            data[40], data[41], data[42]
+        )
```
Comment (cdrini) [inline scripts/partner_batch_imports.py:76]:
```suggestion
        self.length, self.width, self.height = data[40:43]
```

File: scripts/partner_batch_imports.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+        # Assert importable
+        assert self.isbn_13
+        for field in self.REQUIRED_FIELDS:
+            assert getattr(self, field)
+
+    @staticmethod
+    def contributors(data):
+        def make_author(contributor):
```
Comment (cdrini) [inline scripts/partner_batch_imports.py:85]:
```suggestion
        def make_author(name, _, typ):
```

File: scripts/partner_batch_imports.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+            # AU = author
+            # ED = editor
+            return author
+
+        contributors = (
+            (data[21+i*3], data[22+i*3], data[23+i*3]) for i in range(5)
+        )
+
+        # form list of author dicts
+        authors = [make_author(c) for c in contributors if c[0]]
```
Comment (cdrini) [inline scripts/partner_batch_imports.py:100]:
```suggestion
        authors = [make_author(*c) for c in contributors if c[0]]
```

File: scripts/partner_batch_imports.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+        return authors
+
+    def json(self):
+        return dict(
+            (field, getattr(self, field))
+            for field in self.ACTIVE_FIELDS
+            if getattr(self, field)
+        )
```
Comment (cdrini) [inline scripts/partner_batch_imports.py:108]:
```suggestion
        return {
            field: getattr(self, field)
            for field in self.ACTIVE_FIELDS
            if getattr(self, field)
        }
```

**Review by cdrini — COMMENTED** [review]
Code overall looks ok; not doing to a deep dive now that there are tests :)

File: scripts/tests/test_partner_batch_imports.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
+from ..partner_batch_imports import Biblio
+
+csv_row = "USA01961304|0962561851||9780962561856|AC|I|TC||B||Sutra on Upasaka Precepts|The||||||||2006|20060531|Heng-ching, Shih|TR||||||||||||||226|ENG||0.545|22.860|15.240|||||||P|||||||74474||||||27181|USD|30.00||||||||||||||||||||||||||||SUTRAS|BUDDHISM_SACRED BOOKS|||||||||REL007030|REL032000|||||||||HRES|HRG|||||||||RB,BIP,MIR,SYN|1961304|00|9780962561856|67499962||PRN|75422798|||||||BDK America||1||||||||10.1604/9780962561856|91-060120||20060531|||||REL007030||||||"  # noqa: E501
+
+
+class TestBiblio():
+    def test_biblio():
```
Comment (cdrini) [inline scripts/tests/test_partner_batch_imports.py:7]: This should fix pytest. (rename is just a rename :P )

```suggestion
class TestBiblio:
    def test_sample_csv_row(self):
```

File: scripts/partner_batch_imports.py [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+        for field in self.REQUIRED_FIELDS:
+            assert getattr(self, field)
+
+    @staticmethod
+    def contributors(data):
+        def make_author(name, _, typ):
+            author = {'name': contributor[0]}
+            if contributor[2] == 'X':
+                # set corporate contributor
+                author['entity_type'] = 'org'
```
Comment (cdrini) [inline scripts/partner_batch_imports.py:87]: :)

```suggestion
            author = {'name': name}
            if typ == 'X':
                # set corporate contributor
                author['entity_type'] = 'org'
```

File: scripts/partner_batch_imports.py [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+    """Records the last file we began processing and the current line"""
+    with open(logfile, 'w') as fout:
+        fout.write('%s,%s\n' % (fname, line_num))
+
+
+def csv_to_ol_json_item(line):
+    """converts a line to a book item"""
+    b = Biblio(line.strip().split('|'))
+    return {
+        'ia_id': 'isbn:%s' % b.isbn,
```
Comment (cdrini) [inline scripts/partner_batch_imports.py:147]: Oh, should this be `bwb:`? Up above I saw `bwb:` used somewhere.

Comment (mekarpeles) [reply scripts/partner_batch_imports.py:147]: There's a comment; if/when we switch to a prefix like bwb we first need to address the ~3M records that start with `isbn:` so I'd rather this solution be addressed outside of this commit.

File: scripts/partner_batch_imports.py [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+    def __init__(self, data):
+        self.isbn = data[124]
+        self.isbn_13 = [self.isbn]
+        self.title = data[10]
+        self.publish_date = data[20][:4]  # YYYY, YYYYMMDD
+        self.publishers = [data[135]]
+        self.weight = data[39]
+        self.authors = self.contributors(data)
+        self.lc_classifications = data[147]
+        self.number_of_pages = next(
```
Comment (hornc) [inline scripts/partner_batch_imports.py:51]: @mekarpeles this has come up before of https://github.com/internetarchive/openlibrary/issues/4204 , but the `data[36]` field should correspond to the OL text field `pagination` , not the int `number_of_pages`.

From memory, the consensus amongst librarian users was that the `pagination` field is more useful, and generally we only get an integer count of number of pages from the IA scanning process. This regex method looks lossy and and is going to undercount pages in many cases, and lose part of the provided pagination information.

File: scripts/partner_batch_imports.py [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+
+def batch_import(path, batch, batch_size=5000):
+    logfile = os.path.join(path, 'import.log')
+    filenames, offset = load_state(path, logfile)
+
+    for fname in filenames:
+        book_items = []
+        with open(fname, 'r', encoding="ISO-8859-1") as f:
```
Comment (hornc) [inline scripts/partner_batch_imports.py:158]: Are you sure the input file is meant to be ISO-8859-1 encoded?
I'm not sure exactly what file you are using, but I though all partner input files claimed to be UTF-8 encoded.

If there are  ISO-8859-1 characters in there, I'd be concerned those are errors, and assuming the entire file is ISO-8859-1 encoded will break the correctly encoded UTF-8.

File: scripts/partner_batch_imports.py (hunk 17 above)
Comment (cdrini) [reply scripts/partner_batch_imports.py:147]: Oh, `bwb:` was used in the source record field; I think you might've started switching but missed a spot. Sg!

**Review by cdrini — APPROVED** [review]
Code lgtm; leaving it to you to respond to @hornc 's comments. Feel free to merge at your discretion.

File: scripts/partner_batch_imports.py (hunk 19 above)
Comment (mekarpeles) [reply scripts/partner_batch_imports.py:154]:
```suggestion
        with open(fname, 'r') as f:
```

**Pull request merged**